### PR TITLE
Introducing Module Decorator for Enhanced IoC and Context Management

### PR DESCRIPTION
**Pull request by danielmoumeny212**
This pull request introduces module declaration using the Module decorator, implements a basic Inversion of Control (IoC) system as a starting point, and establishes an application context for a Dependency Injection (DI) system.

### CODE REVIEW

**Comment by ItayTheDar** [comment]
@danielmoumeny212 Hi daniel! How are you? this is a true masterpiece and I would like us to have a small meeting to talk about your contribution if you mind?

**Comment by danielmoumeny212** [comment]
Hi there! I appreciate your kind words. I'm doing well, thank you. I'd be
happy to schedule a meeting to discuss my contribution. Please let me know
a convenient time for you.

**Comment by pisipisi** [comment]
I definitely need this PR, thanks so much for your great work @danielmoumeny212

**Comment by danielmoumeny212** [comment]
@pisipisi I'm glad you like it! I've been eagerly awaiting the merge for a while now. Hopefully, it gets merged soon! You can check out the fork with the modifications on my GitHub profile

**Comment by ItayTheDar** [comment]
> @pisipisi I'm glad you like it! I've been eagerly awaiting the merge for a while now. Hopefully, it gets merged soon! You can check out the fork with the modifications on my GitHub profile

@danielmoumeny212 I'm also can't wait for this pr to be merged, this will be a huge milestone for PyNest. could we schedule a meeting this week? maybe tomorrow at 16:00?

**Comment by danielmoumeny212** [comment]
@ItayTheDar of course we can ! Absolutely!Tomorrow at 16:00 works for me.

Regarding the time zone, could you please confirm your current geographical location? This will help me ensure that 16:00 on your end aligns with my availability.

**Comment by danielmoumeny212** [comment]
@ItayTheDar @pisipisi Here's the link to my playground project that allowed me to test the features proposed in the PR! Feel free to clone it and see how it works:  https://github.com/danielmoumeny212/pynest_playground

**Comment by ItayTheDar** [comment]
> @ItayTheDar of course we can ! Absolutely!Tomorrow at 16:00 works for me.
> 
> Regarding the time zone, could you please confirm your current geographical location? This will help me ensure that 16:00 on your end aligns with my availability.

Hi @danielmoumeny212! my timezone is GMT+2. I believe that you are from DRC? so you either in GMT+2 (Kinshasa) or GMT+1 (Lubumbashi)?

**Comment by ItayTheDar** [comment]
> @ItayTheDar @pisipisi Here's the link to my playground project that allowed me to test the features proposed in the PR! Feel free to clone it and see how it works: https://github.com/danielmoumeny212/pynest_playground

This is super cool. I'm looking into it!

**Comment by pisipisi** [comment]
@danielmoumeny212  awesome! I will check it out.

**Comment by danielmoumeny212** [comment]
@ItayTheDar  Gmt + 1 from Congo Brazzaville not DRC !

**Comment by danielmoumeny212** [comment]
Hey @ItayTheDar, have you checked out the pynest playground? If so, what are your thoughts on it?

File: nest/core/pynest_factory.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    ):
+        container.add_module(module)
+
+    @staticmethod
+    def create(main_module: T, **kwargs) -> PyNestApp:
+        container = PyNestContainer()
+        http_server = PyNestFactory._create_server(**kwargs)
+        PyNestFactory._initialize(main_module, container)
+        app = PyNestApp(container, http_server)
+        return app
```
Comment (ItayTheDar) [inline nest/core/pynest_factory.py:22]: return app.get_server()

**Review by ItayTheDar — CHANGES_REQUESTED** [review]
Hi @danielmoumeny212, iv'e told you once, but i have to again, this code is just amazing. for a long time I'm looking for the "spice" to add to PyNest that will make it something more then just a nice wrapping of fastapi, and this pr is really that thing. so thank you for that!

a few notes that are more general:
- I wish that PyNest source code will be covered with documentation and types, to make the usage of it super easy and strait forward
- I also wish that the code will. be self-explanatory, with a meaningful variable names, function and classes
- I want us to think about the directory hirarchrey and where every component is sitting (what belongs to core, what to common and inside that, what sub folders contains what code)

once again, thank you for everything, let me know if there is anything I can do to accelerate this pr

File: nest/common/module.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        return next(
+            (module for module in self.values() if module.id == module_id), None
+        )
+
+    def has(self, token: str):
+        return True if self.get(token) is not None else False
+
+
+class Module:
+    def __init__(self, metatype, container):
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/common/module.py:29]: indeed, the container is of type PyNestContainer instance, while the metatype signifies the class type annotated with the @Module() decorator.

File: nest/common/module.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        self._id = str(uuid.uuid4())
+        self._metatype = metatype
+        self.container = container
+        self._imports = set()
+        self._providers = {}
+        self._injectables = {}
+        self._middlewares = {}
+        self._controllers = {}
+        self._entryProviderKeys = set()
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/common/module.py:38]: Yes, I sometimes find myself a bit confused with naming conventions when transitioning from one programming language to another, particularly from TypeScript to Python

File: nest/common/module.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    @property
+    def id(self):
+        return self._id
+
+    @property
+    def token(self):
+        return self._token
+
+    @token.setter
+    def token(self, token):
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/common/module.py:52]: well noted

File: nest/common/module.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+    def add_provider(self, provider) -> str:
+        self._providers[provider.__name__] = provider
+        return provider.__name__
+
+    def add_controller(self, controller) -> str:
+        self._controllers[controller.__name__] = controller
+        return controller.__name__
+
+    def add_exported_provider(self, provider):
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/common/module.py:121]: Yes, they may or may not be implemented in the future, but let's just  follows the YAGNI (You Aren't Gonna Need It) principle for these .

File: nest/common/route_resolver.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
+from fastapi import FastAPI, APIRouter
+from nest.core.pynest_container import PyNestContainer
+
+
+class RoutesResolver:
+    def __init__(self, container: PyNestContainer, app_ref: FastAPI):
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/common/route_resolver.py:6]: yes we can !
maybe http_adapter!

File: nest/common/route_resolver.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+        self.container = container
+        self.app_ref = app_ref
+
+    def register_routes(self):
+        for module in self.container.modules.values():
+            for controller in module.controllers.values():
+                self.register_route(controller)
+
+    def register_route(self, controller):
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/common/route_resolver.py:15]: yes it'll be better

File: nest/core/decorators/injectable.py [hunk 8]
```diff
@@ -0,0 +1,6 @@
+from nest.common.constants import INJECTABLE_TOKEN
+
+
+def Injectable(cls):
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/core/decorators/injectable.py:4]: Yes, I plan to integrate that into the pull request. It's just that, for now, I haven't incorporated it into the initial version of PyNestContainer. In the first version, with the DiocContainer, injection was done as follows:
@Inject([
 UserService,
 ProductService
])
class UserController:
The @Inject decorator instantiated the container to search for the requested services. If found, it added an attribute like userService that would be equal to an instance of the requested service in the decorator. However, this approach isn't quite right. The decorator should simply annotate the class to specify which service to inject, and then the framework's logic should detect these annotated classes to perform the injections.
I believe we can add this functionality together in the pull request.

File: nest/core/pynest_app_context.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+    def init(self):
+        if self._is_initialized:
+            return self
+
+        self._is_initialized = True
+        return self
+
+    def __init__(
+        self, container: PyNestContainer, context_module: Union[Module, None] = None
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/core/pynest_app_context.py:29]: yes it is actually ! 

 def select_context_module(self): 
       modules = self.container.modules.values()
       self.context_module = next(iter(modules), None)
       
This code snippet from PyNestApplicationContext  selects a module from a container by retrieving all modules, and then setting the context_module attribute to the first module found, or None if there are no modules. The purpose is to establish the current module in the application context.

File: nest/core/pynest_application.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        self.register_routes()
+
+    def use(self, middleware: type, **options: Any) -> None:
+        self.http_server.add_middleware(middleware, **options)
+        return self
+
+    def get_server(self):
+        return self.http_server
+
+    def enable_cors(self, options: Dict[str, Union[str, bool, List[str]]]):
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/core/pynest_application.py:31]: There isn't a significant difference between this function and the direct use of CORSMiddleware capabilities. The main idea behind this implementation is to provide users with the ability to enable CORS by simply passing options without worrying about explicitly passing the reference to the CorsMiddleware class.

File: nest/core/pynest_container.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = super(PyNestContainer, cls).__new__(cls)
+        return cls._instance
+
+    def __init__(self):
+        self.logger = logging.getLogger("pynest")
+
+    _global_modules = set()
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/core/pynest_container.py:30]: Yes, currently, there isn't a functionality for defining a global module. However, the infrastructure is in place to potentially introduce this feature in the future.

File: nest/core/pynest_container.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+        return self._modules[module_key]
+
+    def add_module(self, metaclass):
+        module_factory = self._module_compiler.compile(metaclass)
+        token = module_factory.token
+        if self._modules.has(token):
+            return {"module_ref": self.modules.get(token), "inserted": True}
+        return {"module_ref": self.set_module(module_factory), "inserted": False}
+
+    def add_related_module(self, related_module, token: str):
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/core/pynest_container.py:62]: It refers to the Module class! related_module is a module that gets imported into another module using the import array.

File: nest/core/pynest_container.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+
+    def add_providers(self, providers: List[Any], module_token: str):
+        for provider in providers:
+            self.add_provider(module_token, provider)
+
+    def add_controllers(self, controllers: List[Any], module_token: str):
+        for controller in controllers:
+            self.add_controller(module_token, controller)
+
+    def add_controller(self, token: str, controller):
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/core/pynest_container.py:88]: yeah I totally okey with that

File: nest/core/pynest_container.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+
+    def add_provider(self, token: str, provider):
+        module_ref: Module = self.modules[token]
+        if not provider:
+            raise CircularDependencyException(module_ref.metatype)
+
+        if not module_ref:
+            raise UnknownModuleException()
+        if not hasattr(provider, INJECTABLE_TOKEN):
+            error_message = f"""
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/core/pynest_container.py:137]: yes I could

File: nest/core/pynest_factory.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+class PyNestFactory:
+    @staticmethod
+    def _initialize(
+        module: T, container: PyNestContainer,
+    ):
+        container.add_module(module)
+
+    @staticmethod
+    def create(main_module: T, **kwargs) -> PyNestApp:
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/core/pynest_factory.py:17]: Before using this approach, I initially tried to follow the approach you suggested by making PyNestApp inherit from FastAPI to run it with uvicorn, but it didn't work. So, the only approach that made the code work is what I'm currently proposing. As I mentioned before, I don't want to expose the instance of FastAPI to maintain complete abstraction. Currently, I can still use the FastAPI approach with @http_server.get() by example

File: nest/core/pynest_factory.py (hunk 1 above)
Comment (danielmoumeny212) [reply nest/core/pynest_factory.py:22]: By doing that, we'll still use the FastAPI approach with `@app.get()`, for example.

File: nest/common/route_resolver.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ItayTheDar) [reply nest/common/route_resolver.py:6]: http_adapter is great, i like that it follows the standard of NestJS

File: nest/core/decorators/injectable.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ItayTheDar) [reply nest/core/decorators/injectable.py:4]: we need to come up with a way to minimize the amount of code that the user is writing. like in nestjs, you only needs to use the Injectable decorator on your provider, and then you just inject it to the controller in the constructor - 

```typescript
import { Injectable } from '@nestjs/common';
import { Cat } from './interfaces/cat.interface';

@Injectable()
export class CatsService {
  private readonly cats: Cat[] = [];

  create(cat: Cat) {
    this.cats.push(cat);
  }

  findAll(): Cat[] {
    return this.cats;
  }
}
```

File: nest/core/pynest_app_context.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ItayTheDar) [reply nest/core/pynest_app_context.py:29]: amazing

File: nest/core/pynest_app_context.py [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    ):
+        self.container = container
+        self.context_module: Module = context_module
+        self.logger = logging.getLogger(PyNestApplicationContext.__name__)
+
+    def select_context_module(self):
+        modules = self.container.modules.values()
+        self.context_module = next(iter(modules), None)
+
+    def select(self, module):
```
Comment (ItayTheDar) [inline nest/core/pynest_app_context.py:39]: not sure where this function is being called?

File: nest/core/pynest_container.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ItayTheDar) [reply nest/core/pynest_container.py:66]: when one module imported into another, does that has any meaning? it gives module 'a' the capabilities to use model 'b' providers? as in nestjs? do we have enforcements on this?

File: nest/core/pynest_container.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ItayTheDar) [reply nest/core/pynest_container.py:146]: this is resolved i think

File: nest/core/pynest_factory.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ItayTheDar) [reply nest/core/pynest_factory.py:17]: great answer, let's just keep it this way, you can resolve

File: nest/core/decorators/http_method.py [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+Post: Callable[[Union[str, List[str]]], Callable[..., Any]] = lambda path="/", **kwargs: route("POST", path, **kwargs)
+
+# Decorator for defining a DELETE route with an optional path
+Delete: Callable[[Union[str, List[str]]], Callable[..., Any]] = lambda path="/", **kwargs: route("DELETE", path, **kwargs)
+
+# Decorator for defining a PUT route with an optional path
+Put: Callable[[Union[str, List[str]]], Callable[..., Any]] = lambda path="/", **kwargs: route("PUT", path, **kwargs)
+
+# Decorator for defining a PATCH route with an optional path
+Patch: Callable[[Union[str, List[str]]], Callable[..., Any]] = lambda path="/", **kwargs: route("PATCH", path, **kwargs)
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/core/decorators/http_method.py]: The use cases include: 
     Allowing clients to choose between specific paths based on their needs.
     Triggering different logic within the controller method based on the selected path.

File: nest/core/decorators/injectable.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/core/decorators/injectable.py:4]: Yes, for that, we've achieved dependency injection mechanism! The main goal of the @Module module is to auto-configure dependencies for the IoC to perform dependency injection

File: nest/core/decorators/module.py [hunk 18]
```diff
@@ -0,0 +1,14 @@
+from nest.common.constants import ModuleMetadata
+
+
+class Module:
```
[the remark this answers is not in the document]
Comment (danielmoumeny212) [reply nest/core/decorators/module.py:4]: It's just a enum that keys for module metadata such import, controller etc....

**Comment by ELC** [comment]
Any updates on when this PR will get merged?

**Comment by ItayTheDar** [comment]
Hi @ELC! Sorry for the delay. This pr is missing some very important part of the DI mechanism which is the part of initialize the dependencies and store them in a dictionary. I'm working this days on adding this capability so pynest will offer an actual modular architecture.

**Review by ItayTheDar — APPROVED** [review]
@danielmoumeny212, This Pr is ready to ship into production!
Thank you for your huge effort!

**Pull request merged**